src/validation.cpp

```diff
@@ -5183,6 +5185,294 @@ CChainState& ChainstateManager::InitializeChainstate(CTxMemPool& mempool, const
     return *to_modify;
 }
 
+Optional<std::reference_wrapper<const AssumeutxoData>> ExpectedAssumeutxo(
```

Comment:
This seems like a pretty strange return type, especially since the only call site immediately assigns an `Optional<AssumeutxoData>` from the return value so you're going to make a copy anyway. Why not just return a `std::optional<AssumeutxoData>` or `AssumeutxoData*`?

Also, prefer `std::optional<>` over `Optional<>` (`Optional<>` is just a wrapper for `std::optional<>` and is deprecated for new code).

Reply:
See @sipa's feedback here: https://github.com/bitcoin/bitcoin/pull/19806#discussion_r549864140

Reply:
Should probably still use `std::optional` though.

Reply:
Fixed.

Reply:
I believe sipa's point was that you could avoid a copy by returning a `std::optional<std::reference_wrapper<T>>`. However, you're using that return to copy into a `std::optional<T>` in the only place that this is called:

https://github.com/bitcoin/bitcoin/pull/19806/files#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R5422

so you're not actually saving a copy at all.

In any case, a copy of `AssumeutxoData` isn't expensive (32 byte hash + 4 byte int) and this is an infrequent operation, so there's really not problem just returning a `std::optional<AssumeutxoData>` (or indeed a `const AssumeutxoData*>` - returning a pointer with `nullptr` as a notfound sentinel is a common pattern).

Alternatively, if you want to avoid the copy, just don't assign to `std::optional<AssumeUtxoData>` at the call site:

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index b4206025be..8c95a837f8 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -5419,7 +5419,7 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
     // Assert that the deserialized chainstate contents match the expected assumeutxo value.
 
     int base_height = snapshot_start_block->nHeight;
-    Optional<AssumeutxoData> maybe_au_data = ExpectedAssumeutxo(base_height, ::Params());
+    auto maybe_au_data = ExpectedAssumeutxo(base_height, ::Params());
 
     if (!maybe_au_data) {
         LogPrintf("[snapshot] assumeutxo height in snapshot metadata not recognized " /* Continued */
@@ -5427,7 +5427,7 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
         return false;
     }
 
-    const AssumeutxoData& au_data = maybe_au_data.value();
+    const AssumeutxoData& au_data = maybe_au_data.value().get();
```

I personally find the `value().get()` for fishing the reference out a bit ugly.

Reply:
Gotcha, thanks @jnewbery.